### DISCUSSION TITLE
Associated/Replacement Identifiers for Generic Arguments

### DISCUSSION THREAD

**Discussion by daiplusplus** [dotnet/csharplang#7612] [Language Ideas]
In the (perennial) discussions over _"why can't we use 1-tuples for named-values?"_ I see a common use-case is to attach meaningful names/identifiers to scalars without wanting the overhead of having to define a new type (especially one that holds only a single value - though the recently-introduced `record struct` types do help here) - or having to compromise the design by introducing a 2-tuple just so they can use it like a 1-tuple (e.g. `List<( Int32 userId, Byte _ )> listOfUserIds = ...`).

In my own experience, instead of the above, I often find myself just using `/**/`-comments to add labels to generic type-arguments, for example, for a list of `Int32 userId` values, I'd have `List<Int32/*userId*/> userIds = ...`  or `Dictionary<String/*zipCode*/,String/*countyName*/>`, for example.

...then I realised that the C# language could be extended to permit an identifier or label to be used following a generic type argument and expose that label as a replacement identifier for parameters and member-names (even if it's "faked" by storing the name in an attribute at compile-time; just like how C# already does this with `ValueTuple`).

So if we could do this, then a hypothetical tweaked `List<T>` might look like this:

```
public class List<T t> : IList<T t> // Using `t` as the placeholder for where the (optionall) provided identifier is re-used.
{
    public void Add( T t ) { } // So the `t` here is replaced by the identifier from the actual type argument.

    // etc
}
```

And used like so:

```
List<Int32 userId> listOfUserIds = new();
listOfUserIds.Add( userId: 123 ); // <-- Note how C# allows the name `userId:` to be used as the parameter-name instead of `item:`.
```

This could also work for member names too, if we start-off with [the (fantastic) `OneOf<>` library](https://github.com/mcintyre321/OneOf):

```
OneOf< FileInfo fileOnLocalDisk, Uri azureBlobStorage, Stream other > dataSource = ...
```

...then the members of `dataSource` would have "_virtual-names_" (just like `ValueTuple`) matching those identifiers (can we have auto-PascalCasing yet?) instead of having to use the generic-named `IsT0` `IsT2`, `IsT2`, `TryPickT0`, `TryPickT1` members:

```
if( dataSource.TryPickFileOnLocalDisk( out FileInfo fi, out var or2 ) ) {
      await DoStuffAsync( new FileStream( fi ) );
}
else if( or2.TryPickAzureBlobStorage( out Uri uri, out Stream other ) ) {
      await DoStuffAsync( GetAsync( uri ) );
}
else {
      await DoStuffAsync( other );
}
```

**Comment by HaloFour** [comment]
See: #1883

**Reply by daiplusplus** [reply]
Thanks - I didn't see that earlier discussion when I searched before posting this. Whoops.
